File: 2024/14.py
import time, numpy as np, copy

from utils import readInput, manhattan_dist

# FIXME PROVARE CON DISCRETE FOURIER TRANSFORM

def loadInput(filename: str) -> tuple[list[complex], list[complex], int, int]:
    lines = readInput(filename)
    xmax, ymax = 101, 103

    robots: list[complex] = []
    vels: list[complex] = []
    for l in lines:
        items = l.split()
        robots.append(complex(*map(int, items[0].split("=")[1].split(","))))
        vels.append(complex(*map(int, items[1].split("=")[1].split(","))))
    return robots, vels, xmax, ymax

def quadrants(robots: list[complex], xmax: int, ymax: int) -> list[int]:
    quadrants = [0, 0, 0, 0]
    for r in robots:
        if r.real < xmax//2:
            if r.imag < ymax//2:
                quadrants[0] += 1
            elif r.imag > ymax//2:
                quadrants[1] += 1
        elif r.real > xmax//2:
            if r.imag < ymax//2:
                quadrants[2] += 1
            elif r.imag > ymax//2:
                quadrants[3] += 1
    return quadrants

def part1(robots: list[complex], vels: list[complex], xmax: int, ymax: int):
    robots = copy.deepcopy(robots)
    for i in range(len(robots)):
        npos = complex(robots[i]+vels[i]*100)
        robots[i] = complex(npos.real%xmax, npos.imag%ymax)
    print (f"🎄 Part 1: {np.prod(quadrants(robots, xmax, ymax))}", end='')

def draw(ps: list[complex], xmax: int, ymax: int):
    for y in range(0, ymax):
        for x in range(0, xmax):
            pos = complex(x, y)
            c = ps.count(pos)
            if c != 0:
                print (str(c), end='')
            else:
                print (" ", end='')
        print (" ")

def mean_distance_to_centroid(robots: list[complex], distance=manhattan_dist) -> float:
    centroid = np.mean(robots)
    return np.mean([distance(r, centroid) for r in robots])

def part2_bis(robots: list[complex], vels: list[complex], xmax: int, ymax: int):
    seconds = 10403 # found when the starting position is found again
    minimum = (-1, float("inf"))
    for second in range(1, seconds):
        for i in range(len(robots)):
            npos = complex(robots[i]+vels[i])
            robots[i] = complex(npos.real%xmax, npos.imag%ymax)
        dist = mean_distance_to_centroid(robots)
        if dist < minimum[1]:
            minimum = (second, dist)
    print (f"🎄🎅 Part 2: {minimum}", end='')

def part2(robots: list[complex], vels: list[complex], xmax: int, ymax: int):
    seconds = 10403 # found when the starting position is found again

    stds = (-1, float("inf"), float("inf"))
    for second in range(1, seconds):
        for i in range(len(robots)):
            npos = complex(robots[i]+vels[i])
            robots[i] = complex(npos.real%xmax, npos.imag%ymax)

        std = (np.std([float(r.real) for r in robots]), np.std([float(r.imag) for r in robots]))
        if std[0] <= stds[1] and std[1] <= stds[2]:
            stds = (second, std[0], std[1])

    # for i in range(len(r)):
    #     npos = complex(r[i]+vels[i]*christmas_tree)
    #     r[i] = complex(npos.real%xmax, npos.imag%ymax)
    #draw(r, xmax, ymax)
    print (f"🎄🎅 Part 2: {stds[0]}", end='')

if __name__ == '__main__':
    title = "Day 14: Restroom Redoubt"
    sub = "❄ "*(len(title)//2+2)

    print()
    print(f" {title} ")
    print(sub)
    
    inputs = loadInput("input_14.txt")
    
    t0 = time.time()
    part1(*inputs)
    print (" - {:.6f}s".format(time.time()-t0))
    
    t0 = time.time()
    part2(*inputs)
    print (" - {:.6f}s".format(time.time()-t0))

File: 2024/14.jl
using Statistics

function loadInput(filename::String)
    lines = readlines(filename)
    xmax, ymax = 101, 103
    #xmax, ymax = 11, 7

    robots = []
    vels = []
    for l in lines
        push!(robots, Complex(parse.(Int, split(split(split(l)[1], "=")[2], ","))...))
        push!(vels, Complex(parse.(Int, split(split(split(l)[2], "=")[2], ","))...))
    end
    return robots, vels, xmax, ymax
end
      
function quadrants(robots, xmax, ymax)
    quadrants = [0, 0, 0, 0]
    for r in robots
        if real(r) < xmax÷2
            if imag(r) < ymax÷2
                quadrants[1] += 1
            elseif imag(r) > ymax÷2
                quadrants[2] += 1
            end
        elseif real(r) > xmax÷2
            if imag(r) < ymax÷2
                quadrants[3] += 1
            elseif imag(r) > ymax÷2
                quadrants[4] += 1
            end
        end
    end
    return quadrants
end

function draw(ps, xmax, ymax)
    for y in 0:ymax-1
        for x in 0:xmax-1
            pos = Complex(x, y)
            if pos in ps
                print('X')
            else
                print('.')
            end
        end
        println("")
    end
end

# def mean_distance_to_centroid(robots, distance=manhattan_dist) -> float:
#     centroid = np.mean(robots)
#     return np.mean([distance(r, centroid) for r in robots])

# def part2_bis(robots, vels, xmax, ymax):
#     seconds = 10403 # found when the starting position is found again
#     minimum = (-1, float("inf"))
#     for second in range(1, seconds):
#         for i in range(len(robots)):
#             npos = complex(robots[i]+vels[i])
#             robots[i] = complex(npos.real%xmax, npos.imag%ymax)
#         dist = mean_distance_to_centroid(robots)
#         if dist < minimum[1]:
#             minimum = (second, dist)
#     print (f"🎄🎅 Part 2: {minimum}", end='')

function python_mod(x::Int, y::Int)
    if y == 0
        error("Division by zero")
    end
    r = x % y
    return r + (r < 0 && y > 0 ? y : 0)
end

function part1(robots, vels, xmax, ymax)
    for i in eachindex(robots)
        npos = robots[i] + vels[i]*100
        robots[i] = Complex(python_mod(Int(real(npos)), xmax), python_mod(Int(imag(npos)), ymax))
    end
    #draw(robots, xmax, ymax)
    μ = prod(quadrants(robots, xmax, ymax))
    println("🎄 Part 1:  $(μ)")
end

function part2(robots, vels, xmax, ymax)
    seconds = 10403 # found when the starting position is found again
    stds = (-1, Inf, Inf) 
    
    for second in 1:seconds
        for i in 1:length(robots)
            npos = robots[i] + vels[i]
            robots[i] = Complex(python_mod(Int(real(npos)), xmax), python_mod(Int(imag(npos)), ymax))
        end
    
        std_x = std(real.(robots))
        std_y = std(imag.(robots))
    
        if std_x <= stds[2] && std_y <= stds[3]
            stds = (second, std_x, std_y)
        end
    end
    println("🎄🎅 Part 2: $(stds[1])")    
end

function main()
    title = "Day 14: Restroom Redoubt"
    sub = "❄ "^(length(title)÷2 + 2)
    
    println()
    println(" $title ")
    println(sub)
    
    inputs = loadInput("input_14.txt")
    @time part1(inputs...)
    inputs = loadInput("input_14.txt")
    @time part2(inputs...)
end

main()